### ISSUE TITLE
Migrate to Toggl API v9 and Reports API v3

### ISSUE THREAD

**Issue by mcndt** [saintedlama/toggl-client#23]
Toggl has once again decided that they need to update their public API: https://developers.track.toggl.com/docs/index.html

> This documentation focuses on current API v9 and the old v8 version will soon be deprecated. The documentation is already extended to [Webhooks API](https://developers.track.toggl.com/docs/webhooks_start) and will soon be extended to missing services like Reports API v3.

It will probably be a while until the v8 and reports v2 APIs are deprecated, but this issue can serve as a discussion on the migration.

**Comment by beauraines** [comment]
I saw that the v8 endpoint is to be turned off at the end of the month, so that pushed me to investigate a little.

[Announcement](https://toggl.com/blog/toggl-track-api-v9) 

[Migration guide](https://developers.track.toggl.com/docs/additional/v8_migration_guide)

Although they say

> Keep in mind that replacing the endpoints is not enough. Please make sure you follow the directions in the documentation to check how the new endpoints work.

This change allowed all the smoke tests to still pass.

```diff
diff --git a/lib/client.js b/lib/client.js
index 499cbbf..0ea48be 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -50,7 +50,7 @@ class TogglClient {
     this.user = new User(this);
 
     this.httpClient = got.extend({
-      prefixUrl: 'https://api.track.toggl.com/api/v8',
+      prefixUrl: 'https://api.track.toggl.com/api/v9',
       username: this.options.apiToken,
       password: 'api_token',
       throwHttpErrors: false,

```

```bash
npm test

> toggl-client@1.1.0 test
> mocha



  smoke test
    ✓ should list workspaces (250ms)
    ✓ should list projects in a workspace (462ms)
    ✓ should get a details report (469ms)
    ✓ should get a user (110ms)
    - should get a new API token
    ✓ should throw an error if current password not supplied
    ✓ should throw an error if time of day format is invalid
    ✓ should throw an error if date format is invalid
    - should generate time entries


  7 passing (1s)
  2 pending
```

And looking at the output of the tests (I swapped `debug` for `console.log` as I don't really know mocha or unit testing very well), the output visually looked like what I was expecting.

Now, I've not updated my client to the ESM version yet and am running off of a fork just before that change was made, but I'll see what happens to on the client end.

It seems that the API calls are all being made to the `workspace` endpoints, which is what v9 seemed to move to, so this ay be mostly v9 compliant already!!

**Comment by beauraines** [comment]
After poking around with my client to make the call, some of the response attributes have changed.

Looking only at `/me`, the user information is no longer nested in a `data` property and many properties have been removed. Maybe the `toggl-client` callling the API doesn't need to change much, but the changes are on the clients, like mine that uses @saintedlama 's package.

## v8 

```javascript
{
  since: 1676835290,
  data: {
    id: 'OBFUSCATED',
    api_token: 'OBFUSCATED',
    default_wid: 'OBFUSCATED',
    email: 'beau.raines@gmail.com',
    fullname: 'Beau Raines',
    jquery_timeofday_format: 'h:i A',
    jquery_date_format: 'm/d/Y',
    timeofday_format: 'h:mm A',
    date_format: 'MM/DD/YYYY',
    store_start_and_stop_time: true,
    beginning_of_week: 0,
    language: 'en_US',
    image_url: 'https://assets.track.toggl.com/avatars/913dfa76c01d730711eb00a1c3e6d57c.jpg',
    at: '2022-05-08T01:37:13+00:00',
    created_at: '2013-07-13T16:23:55+00:00',
    record_timeline: false,
    should_upgrade: true,
    timezone: 'America/Los_Angeles',
    openid_enabled: true,
    openid_email: 'beau.raines@gmail.com',
    send_product_emails: true,
    send_weekly_report: true,
    send_timer_notifications: true,
    invitation: {},
    workspaces: [ [Object] ],
    used_next: true,
    duration_format: 'classic'
  }
}
```

## v9

```javascript
{
  id: 'OBFUSCATED',
  api_token: 'OBFUSCATED',
  email: 'beau.raines@gmail.com',
  fullname: 'Beau Raines',
  timezone: 'America/Los_Angeles',
  default_workspace_id: 'OBFUSCATED',
  beginning_of_week: 0,
  image_url: 'https://assets.track.toggl.com/avatars/913dfa76c01d730711eb00a1c3e6d57c.jpg',
  created_at: '2013-07-13T16:23:55.630123Z',
  updated_at: '2022-05-08T01:37:13.126904Z',
  openid_email: 'beau.raines@gmail.com',
  openid_enabled: true,
  country_id: 235,
  at: '2023-02-19T19:35:38.009419Z',
  intercom_hash: 'OBFUSCATED',
  oauth_providers: [ 'google' ],
  has_password: true
}
```

**Comment by saintedlama** [comment]
@beauraines I think the change for the `me` endpoint does make a lot of sense not to have that nested `data` object in the response. I think it could make sense to make you a collaborator on this project as I'm not using toggl track currently.

Btw: To get debug output written to the console just set an environment variable `DEBUG` to 'toggl-client-tests' .

**Comment by beauraines** [comment]
@saintedlama Thanks for the mini-PR, for the tests. I saw that before I went to bed last night and it was exactly the clue I needed to update the tests. Thank you. And thanks for the `DEBUG` environment variable tip!

Yesterday afternoon, I made _some_ changes to the API to update it. The pattern seems to be 

1. update the endpoint 
2. no need for the `mapData` function

I'll push a mini PR shortly so you can see the kinds of things I found.

I would love to be a collaborator on toggl-client I've rekindled my love for the CLI and I use Toggl Track almost solely through the CLI. Anything I need to do on the web app, I just haven't built it into my client yet.

**Comment by mcndt** [comment]
For what it's worth, my Obsidian integration for Toggl (https://github.com/mcndt/obsidian-toggl-integration) also depends on this project (about a thousand active users) so I'm quite invested in the migration to V9 as well. Let me know if I can help review changes.

**Comment by beauraines** [comment]
@mcndt Can you confirm the `toggl-client` methods you use?

I went over your repo briefly and it looks like you're using.

1. workspaces.list
2. workspaces.projects
3. workspaces.tags (not overlapping)
4. reports.details (not overlapping)
5. reports.summary (not overlapping)

Does that look right? Maybe we can prioritize these along with the ones that my client is using. There is some overlap. I'm using

1. client.reports.weekly
2. client.timeEntries.current
3. client.timeEntries.list
4. client.timeEntries.stop
5. client.timeEntries.update
7. client.user.current PR #28 
8. client.workspaces.list
9. client.workspaces.projects

I also noticed you're using a fork of the `toggl-client`? What features have you added that weren't in @saintedlama 's?

I think it may make sense to just address these methods _first_, because we may only have 10 days left and then come back to the others in each class. If you concur, I'll create an issue for each of the method updates and start chipping away at them. 

I think I'm going to use `feature/switch-to-v9-endpoint` branch as the develop branch for the v9 and that can be merged into main when ready.

Something that is going to trip me up is that I don't get ESM and the command line parser `yargs` I use don't nicely support it yet.

**Comment by beauraines** [comment]
here is my proposed priority order 

client.timeEntries.current PR #28 (toggl-cli-node)
client.timeEntries.list PR #28 (toggl-cli-node)

1. client.user.current (toggl-cli-node)
2. client.workspaces.list (common)
3. client.workspaces.projects (common)
4. client.timeEntries.create  (toggl-cli-node)
5. client.timeEntries.stop (toggl-cli-node)
6. workspaces.tags (obsidian)
7. reports.details (obsidian)
8. reports.summary (obsidian)
9. client.reports.weekly (toggl-cli-node)
10. client.timeEntries.update (toggl-cli-node)

That should keep both of our projects running and then we can iterate on the remaining functions, one class at a time...

**Comment by beauraines** [comment]
I took a break and got a fresh start and did

- [updates workspaces.list to me/workspaces](https://github.com/saintedlama/toggl-client/pull/28/commits/e6ce38924ae7c8cafec4a64641e4717007fe2ca0)
[e6ce389](https://github.com/saintedlama/toggl-client/pull/28/commits/e6ce38924ae7c8cafec4a64641e4717007fe2ca0)

- [Updates time entry create and start](https://github.com/saintedlama/toggl-client/pull/28/commits/3353e86f7e293dfe901523767fc1c62a38cc91ca) 
[3353e86](https://github.com/saintedlama/toggl-client/pull/28/commits/3353e86f7e293dfe901523767fc1c62a38cc91ca)

in that same PR. I noted that the v9 API is expecting a less nested payload than before and it is much stricter about datatype. In my testing, I was passing in `workspace_id` as a string and it was being rejected because the new API expects it to be a number.

Now to watch the next episode of Last of Us.

**Comment by beauraines** [comment]
Why I needed to take a break... I got flummoxed. The Toggl documentation doesn't refer to it, but it seems that there is a `GET workspaces` endpoint. And I got confused as to what I was testing and why it was working on the OLD endpoint.  I finally figured it out and I did switch to the documented `GET me/workspaces`

**Comment by beauraines** [comment]
I started on the `reports.details()` this morning, but I'm running into a some problems with the Reports API version change. I need to switch gears for the day but I'll come back to it in the afternoon.

**Comment by beauraines** [comment]
I fought with the `report.details()` but I won. The report details is no longer assumes a start date, but one is required.

https://developers.track.toggl.com/docs/reports/detailed_reports/index.html#post-search-time-entries

I'm not sure if it makes sense to define this in `toggl-client` or in the actual packages that use it. If we don't define it in `toggl-client`, I guess we need to define some default. Does it make sense that if you _don't_ specify anything in your client that it starts 9 days ago, which I think was the way the previous version functioned? Or should we make some opinion about starting at the first of the month, regardless of the month.

In my client, I only use the current week's worth of data, but I could see some reason to go further. 

@mcndt How does your client use the report data - how far back is the "typical" use case?

**Comment by beauraines** [comment]
Report details and report weekly DONE 

@mcndt If you can check how your app works against the details method, that would be cool.  **Important** changes

1. you must include a `start_date` in the params for `reports.details()`
2. The response is flatter than before, it is now an array of projects with the details in the `time_entries`, when the response was an object before.

@saintedlama or @mcndt  I'm not very mocha fluent. Can you help me with this assertion? I updated the `reports.details()` to throw an error if no start date was passed in, but I'm not sure how to unit test that.

https://github.com/saintedlama/toggl-client/blob/a98427e986c584027d6d20a36b7b00bfa33aa051/test/smoke-test.js#L53-L61

That's all for tonight. I'm making faster work of this than I expected!

**Comment by beauraines** [comment]
Okay, I did reports.summary() too. In theory, that should be everything you need for Obsidian...

@mcndt  The response looks a little difference and like the details the request must include the `start_date`

```json
{
  groups: [
    { id: 4599739, sub_groups: [Array] },
    { id: 174558624, sub_groups: [Array] },
    { id: 175103503, sub_groups: [Array] },
    { id: 186555135, sub_groups: [Array] }
  ],
  page: undefined,
  hasNextPage: false
}
```

**Comment by mcndt** [comment]
Hi guys, great work so far, I'm following along but haven't had time to look in detail. I blocked off some time tonight to reply to your questions and join the migration effort.

**Comment by beauraines** [comment]
Okay. #28  has all the  core functions for my client and @mcndt 's client

There are two unit tests I need some help with and there are some merge conflicts that I'd much rather resolve locally rather than through the webui, but I need to go make dinner now.

**Comment by mcndt** [comment]
> [@mcndt](https://github.com/mcndt) Can you confirm the `toggl-client` methods you use?

```ts
// list of calls I use
workspaces.list()
workspaces.projects()
workspaces.tags()

reports.details()
reports.summary()
reports.detailsAll()

timeEntries.start()
timeEntries.stop()
timeEntries.current()
```

> I also noticed you're using a fork of the `toggl-client`? What features have you added that weren't in [@saintedlama](https://github.com/saintedlama) 's?

I modified three things:

1. I added an [option to set HTTP headers](https://github.com/mcndt/toggl-client/commit/0f5a9604aba06e7bb4428248f95b636ef796dae6). At one point the polling rate of my client was too high, and I was asked by the Toggl team to identify users of my Obsidian extension with a [custom user-agent](https://github.com/mcndt/obsidian-toggl-integration/blob/master/lib/toggl/TogglClient.ts#L5-L9).
2. I added the option to [use Obsidian’s native HTTP client](https://github.com/mcndt/toggl-client/commit/0fc9dc9771fd4a9adacad07ca617d65e152b83c2) instead of using `got`.
3. I added an [exponential back-off when hitting rate limits](https://github.com/mcndt/toggl-client/commit/8d4d55c8ef7e9fb4855cbbf2575e151bb4249e6e) (because their reports API v2 was quite limited I had to collect stats in pages, which can be many requests when querying data that spans up to a year)

> [@mcndt](https://github.com/mcndt) How does your client use the report data - how far back is the "typical" use case?

As far as the API allows :) My client allows users to define SQL-like queries on their Toggl data, so in theory anything goes. For example:

````
```toggl
SUMMARY FROM 2021-09-21 TO 2021-12-23
INCLUDE CLIENTS 
    "Master’s Thesis", "Computer Graphics", 
    "Queueing Analysis", "AI Research Seminar", 
    "Internship"
SORT DESC
TITLE "Fall Semester 2021-2022"
```
````

This query then fetches all the pages of the detail reports required to render the following chart:

<img width="700" alt="image" src="https://user-images.githubusercontent.com/23149353/221284775-0d521b72-2db9-4ead-8078-618ea9f3f830.png">

**Comment by mcndt** [comment]
My god, the V9 documentation is atrocious...

**Comment by beauraines** [comment]
@mcndt I didn't update `reports.detailsAll()` I missed that one when I went over your code. I'll work on that this weekend.

> My god, the V9 documentation is atrocious...

I agree. While it looks pretty, I could not find an example of a request. It took trial and error to figure out that the un-nested the response data. There are lots more required parameters and they don't stand out.

>>  How does your client use the report data - how far back is the "typical" use case?
>
> As far as the API allows :) My client allows users to define SQL-like queries on their Toggl data, so in theory anything goes. 

I asked this question too soon. I had stumbled upon one of the `start_date` required endpoints and was going to have the `toggl-client` provide a default value so I wanted to know what was reasonable. But then I came to the conclusion that the API shouldn't be opinionated about that and added the check for the required (or at least what I could figure out) parameters.

Though, in the "everything goes" scenario, I was trying one of the report end points (I think or may the entries list), and I accidentally put in a date from 2020 and it gave me a too far back error. I can't remember which, but I didn't add any max date validation, just that the date was included.

And what you built for Obsidian is pretty cool! I love Markdown-like use cases were plain text renders into something cool. That visualization is quite polished and looks very nice.

**Comment by beauraines** [comment]
@mcndt I did update `reports.DetailsAll()` and it is one of those that now _requires_ a start date. I think the old version used to give you 7 days ago, if you didn't provide a date...

**Comment by beauraines** [comment]
@mcndt A few tips from when I updated my app to use the new v9 API

1. The toggl  API is much more restrictive on types. project and workspace ids _must_ be numbers.
2. Durations returned by the weekly report now are seconds when they were milliseconds previously.

How is your upgrade coming?

**Comment by mcndt** [comment]
@beauraines As a matter of fact I pushed out the update last night! I ended up using some different endpoints than before because their summary API is now a lot faster (as in generate a 1 year summary in under a second) whereas previously I had to fetch many pages of data to get the same result.

There are some changes I had to make to pagination of detail reports, I'll make an upstream PR for it sometime this week.
